src/crystal/system/unix/pthread.cr

```diff
@@ -9,20 +9,35 @@ module Crystal::System::Thread
     @system_handle
   end
 
+  protected setter system_handle
+
   private def init_handle
-    # NOTE: the thread may start before `pthread_create` returns, so
-    # `@system_handle` must be set as soon as possible; we cannot use a separate
-    # handle and assign it to `@system_handle`, which would have been too late
+    # NOTE: `@system_handle` needs to be set here too, not just in
+    # `.thread_proc`, since the current thread might progress first; the value
+    # of `LibC.pthread_self` inside the new thread must be equal to this
+    # `@system_handle` after `pthread_create` returns
     ret = GC.pthread_create(
       thread: pointerof(@system_handle),
       attr: Pointer(LibC::PthreadAttrT).null,
-      start: ->(data : Void*) { data.as(::Thread).start; Pointer(Void).null },
+      start: ->Thread.thread_proc(Void*),
       arg: self.as(Void*),
     )
 
     raise RuntimeError.from_os_error("pthread_create", Errno.new(ret)) unless ret == 0
   end
 
+  def self.thread_proc(data : Void*) : Void*
```

Comment:
**question:** What's the motivation for turning this into a method? I understand it's still exclusively used as a proc. So why not keep it as a proc iteral then?

Reply:
IMO the body is now large enough that it should be moved into its own method, just like the Win32 implementation

Reply:
Ah, thanks for the precedent reference. I hadn't notice that.
It's fine to do it that way. I'd probably prefer to rather have procs contained than exposed, but that's a style question which shouldn't matter here.